cmd/zfs/zfs_main.c

```diff
@@ -4287,9 +4307,28 @@ zfs_do_send(int argc, char **argv)
 			}
 		}
 
-		zhp = zfs_open(g_zfs, argv[0], ZFS_TYPE_DATASET);
-		if (zhp == NULL)
-			return (1);
+		/*
+		 * If this is a partial send, the partial data will either
+		 * live in the dataset itself (if the dataset did not
+		 * previously exist) or it will live in the special '%recv'
+		 * clone. To figure out which is correct we simply check both.
+		 */
+		if (flags.partial) {
+			(void) snprintf(tobuf, sizeof (tobuf), "%s/%%recv",
+			    argv[0]);
+			if (zfs_dataset_exists(g_zfs, tobuf,
+			    ZFS_TYPE_DATASET)) {
+				zhp = zfs_open(g_zfs, tobuf, ZFS_TYPE_DATASET);
+				if (zhp == NULL)
+					return (1);
+			}
+		}
+
+		if (zhp == NULL) {
+			zhp = zfs_open(g_zfs, argv[0], ZFS_TYPE_DATASET);
+			if (zhp == NULL)
+				return (1);
+		}
 
```

Comment:
I think the above code could be refactored to only call zfs_open once.
```
diff --git a/cmd/zfs/zfs_main.c b/cmd/zfs/zfs_main.c
index ed52997..09e425b 100644
--- a/cmd/zfs/zfs_main.c
+++ b/cmd/zfs/zfs_main.c
@@ -4099,7 +4099,7 @@ zfs_do_send(int argc, char **argv)
        char *toname = NULL;
        char *resume_token = NULL;
        char *cp;
-       zfs_handle_t *zhp = NULL;
+       zfs_handle_t *zhp;
        sendflags_t flags = { 0 };
        int c, err;
        nvlist_t *dbgnv = NULL;
@@ -4313,22 +4313,19 @@ zfs_do_send(int argc, char **argv)
                 * previously exist) or it will live in the special '%recv'
                 * clone. To figure out which is correct we simply check both.
                 */
+               char *to_open = argv[0];
                if (flags.partial) {
                        (void) snprintf(tobuf, sizeof (tobuf), "%s/%%recv",
                            argv[0]);
                        if (zfs_dataset_exists(g_zfs, tobuf,
                            ZFS_TYPE_DATASET)) {
-                               zhp = zfs_open(g_zfs, tobuf, ZFS_TYPE_DATASET);
-                               if (zhp == NULL)
-                                       return (1);
+                               to_open = tobuf;
                        }
                }
 
-               if (zhp == NULL) {
-                       zhp = zfs_open(g_zfs, argv[0], ZFS_TYPE_DATASET);
-                       if (zhp == NULL)
-                               return (1);
-               }
+               zhp = zfs_open(g_zfs, to_open, ZFS_TYPE_DATASET);
+               if (zhp == NULL)
+                       return (1);
 
                if (fromname != NULL && (strchr(fromname, '#') == NULL &&
                    strchr(fromname, '@') == NULL)) {
```

Reply:
done